### PR TITLE
Build time optimization with include-what-you-use

### PR DESCRIPTION
**Pull request by noseglasses**
# Build time optimization

This is a follow up to discussion https://github.com/keyboardio/Kaleidoscope/issues/736.

## TL;DR

This PR reduces firmware build times by ~20%.

## Motivation

Firmware builds take longer than necessary. That's especially annoying if frequent rebuilds are required during development or if one needs to wait till the smoke-examples test bench runs through before submitting a PR.

## Reasons for build time accumulation

Kaleidoscope has one central header `Kaleidoscope.h` that is meant to simplify handling of the firmware for users. This header pulls in a lot of stuff and is unfortunately included in many places of the firmware core. Thus, many compilation units include stuff that is not necessary for them to build. Also, `Kaleidoscope.h` includes stuff that is only required in the sketch, like the mostly templated infrastructure of the plugin API and transient LED modes, just to name two of them.

### Conditionally excluding parts of the code

A first attempt was to reduce build times by introducing a wrapper compiler that command-line defines a macro that can be used to distinguish the sketch compilation unit from all others. This was meant to suppress infrastructure stuff for all other compilation units and only make it available to the sketch.
This comes with two major drawbacks, namely the wrapper compiler and the burden to figure out what to suppress and how to do it. Especially as this must be done in a way that it remains maintainable.
After experimenting with this idea for a while, I gave up.

#### Where/what to clean up

A second attempt to reduce build time was far more successful. Instead of excluding parts of the code, I cleaned up the whole include management - partially automatized - which lead to a build time decrease of approx. 20%.

Before I go into detail how that works, I would like to draw a distinction between different types of headers that can be found in Kaleidoscope.

* sketch-only headers (headers that are exclusively meant to be included by the sketch)
* library internal headers (included within Kaleidoscope and not meant to be included directly in the sketch)
* private headers (mostly in kaleidoscope_internal)

Next, some rules that are commonly regarded as good practice in C/C++ programming and if respected keep build times low and also re-build times (due to less unnecessary include dependencies).

* headers should only include other headers in very rare cases 
* forward declare whenever possible (therefore, prefer call by reference to call by value) to avoid unnecessary includes
* always try to include headers directly, not via other headers (prevents unnecessary dependencies and improves transparency (what lives where))
* name headers in a way that they reflect the content (simplifies include management and finding appropriate headers)

There are sometimes, for good reasons, exceptions to some of these rules. A typical example are the service-headers that many widely used C++ libraries support. Those are often named by the project and pull in a lot of stuff, sometimes all features, a library offers. This is meant to spare users finding out what symbols are defined where. The downside of this approach is that such headers are really costly on the including side, especially if only few features are required but the compiler sees and parses all of them.
This becomes even worse if such service-headers are included by other central headers of the own project which are in turn included by many files. Such service-headers could easily cause an explosion of build times when used carelessly. That's why it is recommended to use feature-headers instead and preferably include them only in source files (.cpp) and never in headers. This is especially important for the libraries that depend on other libraries.

In Kaleidoscope, an example for such a central service-header is `Kaleidoscope.h`. Also most of the headers directly living in `src`, that originally were the Arduino-library headers of individual plugins before the mono-repo fusion. All those headers are there for good reasons. The trade off between usability and build times should clearly be on the usability side. That's why no attempt is made to speed up sketch build time. 

Things are very much different for most other headers and source files. For those the good-practice rules of inclusion should be respected. Any other compilation units apart from the sketch should really build as fast as possible.

#### Include hell

Unfortunately, a include hell in Kaleidoscope comes from the fact that many compilation units depend on class `Kaleidoscope` that is defined in header `kaleidoscope/Kaleidoscope.h`. The big problem is that this header pulls in a huge lot of other stuff. That's why currently every compilation unit pulls in great parts of Kaleidoscope's headers without actually needing them to build.

#### Refactoring Kaleidoscope.h

The first step in refactoring was to move the definition of class `Kaleidoscope` to a separate header. I called it `Kaleidoscope_Class.h` due to a lack of better ideas. Both the central header `Kaleidoscope.h` but also the header of class `Kaleidoscope` deserve to be named `Kaleidoscope.h`. To remain backward compatible, we definitely have to stick with the name for the central header. Maybe we might consider ?another name for class `Kaleidoscope`? Just an idea that's fairly off topic.

After this change, all compilation units that only need class `Kaleidoscope` can include `Kaleidoscope_Class.h`. This already removes a lot of unnecessary dependencies.

#### Include-what-you-use

After factoring out class `Kaleidoscope` there are still a lot of unnecessary includes left as not all `#include "kaleidoscope/Kaleidoscope.h"` can be replaced by `#include "kaleidoscope/Kaleidoscope_Class.h"`. That's because a great number of such includes actually mean to pull in other stuff that is indirectly included by `Kaleidoscope.h`, e.g. `layer.h` or `hooks.h`. For authors it was probably tempting in the past to just include `Kaleidoscope.h` instead of finding out where a symbol or macro actually lives, knowing that one include would solve all problems.

Fortunately, there is a free tool named [include-what-you-use](https://include-what-you-use.org/) (IWYU), that helps to deal with such problems. It's clang based analyser that can serve as a drop-in replacement for the gcc or clang compiler.

IWYU compiles source files (and also headers if necessary) and provides suggestions about where to remove unnecessary includes or which indirect includes could be replaced by direct ones. Moreover, it also helps with replacing some includes by forward declarations. It is obvious that to provide suggestions about the latter, it needs to analyse the entire AST, which explains why it is based on clang in the first place.

Unfortunately, IWYU cannot be used to analyse Kaleidoscope directly. That's because IWYU cannot replace avr-gcc. There's just no avr-clang or avr-IWYU that could deal with all the Arduino/AVR specific compiler flags.

Luckily, we have virtual builds that already work with x86 compilers. With few changes to the code, Kaleidoscope also builds with latest clang.

I let IWYU run on all headers in `src/kaleidoscope` and `src/kaleidoscope_internal` and replaced include sections with what IWYU suggested. Only in rare cases I had to manually replace some of the suggested standard headers with those available with Arduino. That's because of the virtual build.

The result is a 20% build time reduction (determined by looking at the build time of the stock firmware).

I suggest to repeat this IWYU-procedure once in a while to keep the amount of unnecessary includes low.
FWIW, IWYU comes with a auto-replacement script but that cannot be used (with AVR builds) due to the need to manually replace specific header suggestions.

## Changes

This PR comes with two commits. The first one extracts header `Kaleidoscope_Class.h` from `Kaleidoscope.h`.  The second commit contains the IWYU optimization.

These changes mostly affect include directives being removed, changed or replaced by forward declarations. Any other changes part of the first commit were necessary to enable factoring out class `Kaleidoscope`.

### CODE REVIEW

**Comment by noseglasses** [comment]
I forgot to mention/note to myself: A next step would be to replace any remaining includes of headers that live in `src` with matching headers in `src/kaleidoscope` or `src/kaleidoscope_internal`, there probably aren't many of such occurrences, though.

For aesthetical reasons, any library headers should be included using `"header.h"` instead of `<header.h>` which would be easy to adapt for the entire project using regex. On the other hand I encountered some occurrences of standard headers being included as e.g. "stdint.h" instead of <stdint.h>. Both issues were only addressed by this PR when affected by changes suggested by IWYU.

**Comment by obra** [comment]
This is super-cool.  

I want to think a bit about the "Kaleidoscope_Class" name. I feel like we can find something better.

Also, I'd love to have the PR's writeup as a file inside doc/  - It's a useful explanation of how not to screw this up :) (Over time, hopefully it'll be maintained and improved, but at the least, it'd be sad to lose the text in pull request history.

**Comment by noseglasses** [comment]
* rebased after recent merges
* moved PR's writeup to `doc/build_time_reduction.md`

It just occurred to me that the code of the Dygma Raise was not covered by this optimization because it was merged after my kaleidoscope-virtual sandbox was checked out. I anyway could not test the Raise as the device is missing in `examples/Devices`.

Optimizations are pretty straight forward. Most importantly, a remaining include of `Kaleidoscope.h` must be replaced by directly including other dependee headers.

**Comment by algernon** [comment]
I'll deal with Raise manually. There's no example for it yet, because we can't compile it with `kaleidoscope-builder` (and there are currently no plans to change that), and the `Kaleidoscope-Bundle-Dygma` bundle isn't very up-to-date either. At the moment, the Raise compiles on my box, and I need to push the build instructions and whatnot up.

**Comment by noseglasses** [comment]
> Instead of kaleidoscope/Kaleidoscope_Class.h, I'd suggest either kaleidoscope_internal/Kaleidoscope.h, or kaleidoscope/Kaleidoscope_.h The latter matches the class name, after all. I think the first one is less confusing, but we don't really have a good policy about what goes into internal and what doesn't.

The Kaleidoscope_ class is safe to be internal, I think. The Kaleidoscope object is public though. But using the object (outside of Kaleidoscope itself) is always going to be through the main Kaleidoscope.h header, so the class being in internal should be ok, in my opinion.

I prefer the file name `Kaleidoscope_.h`. I am fine with both locations to put it. It can be confusing to have more than one file with the same name in multiple directories unless its appears many times in a structured way. We already have two `device.h` headers :| Naming the file like the class makes sense to me. Let's see what @obra has to say.

**Comment by obra** [comment]
I think kaleidoscope_internal/Kaleidoscope_.h makes more sense than kaleidoscope/Kaleidoscope_.h or kalediscope_internal/Kaleidoscope.h

>  but we don't really have a good policy about what goes into internal and what doesn't.

If we were starting from scratch, I'd say that things inside kaleidoscope_internal are things that should never be directly included from a user's sketch and anything inside kaleidoscope/ is effectively part of the public API. But that ship has sailed.

**Comment by obra** [comment]
Now that I'm trying the change I proposed, it looks -wrong-. Reevaluating.

**Comment by obra** [comment]
Because the thing that this provides to developers isn't really that Kaleidoscope_ class, but the instance of it, naming it as Kaleidoscope_.h doesn't actually feel right.

Similarly, since it gets used -everywhere-, the kaleidoscope_internal/ directory feels kind of wrong.

**Comment by obra** [comment]
Part of me wants to rename the thing we call 'Kaleidscope' inside namespace kaleidoscope {} to 
something like 'Core' or 'Runtime'  inside namepace kaleidoscope {}

**Comment by noseglasses** [comment]
> Part of me wants to rename the thing we call 'Kaleidscope' inside namespace kaleidoscope {} to
something like 'Core' or 'Runtime' inside namepace kaleidoscope {}

I thought so, too, but hesitated to propose that because of the impact of aname change.

**Comment by obra** [comment]
Since this change is going to touch all those files anyway... (And we -can-
provide backcompat for the name)

Hrm.

I just realized that we -also- have the toplevel "Kaleidoscope.h" file.

Three "Kaleidoscope.h" files mean that we're naming things wrong.

But maybe we should be using that toplevel Kaleidoscope.h for the thing
that user sketches, etc should be including.
ᐧ

**Comment by obra** [comment]
Testing "time make" inside Kaleidoscope on hot and cold builds with a clean ccache and none of the /tmp cache for builds:

# PR 

## First build:

real	3m6.134s
user	3m23.078s
sys	1m17.298s

## Second build (no changes):

real	2m13.822s
user	1m38.910s
sys	0m52.153s


# Master

## First build:

real	3m11.500s
user	4m3.563s
sys	1m19.521s


## Second build (no changes):

real	2m45.180s
user	2m0.359s
sys	1m4.446s

# PR with Kaleidoscope_Class change reverted

## First build

real	3m22.611s
user	4m8.697s
sys	1m24.264s

## Second build (no changes):

real	2m44.939s
user	2m0.461s
sys	1m4.371s

--- it looks like almost all of the win we're getting is from pulling out that core class.

**Comment by obra** [comment]
...and getting rid of kaleidoscope/Kaleidoscope.h, splitting the
functionality to push the user-sketch stuff up to src/Kaleidoscope.h and
the special class to src/kaleidoscope/Runtime.h


ᐧ

**Comment by obra** [comment]
Since IWYU seems to be a tiny percentage of the improvement we're seeing,
let's use this PR to sort out the teasing apart of the "Kaleidoscope.h"
file and then handle IWYU afterward as a separate PR.

(Both are going to stomp on a lot of files.)

My -current- proposal is that the stuff we currently stick in
src/kaleidoscope/Kaleidoscope.h but intend as "what you need in a sketch"
should live in src/Kaleidoscope.h  and the class currently called
Kaleidoscope inside the kaleidoscope namespace be renamed to "Runtime"
inside the kaleidoscope namespace, with back-compat left around for a bit.

I -believe- that @algernon is roughly in agreement on this
plan.  @noseglasses - I'd love to have your opinion.

> ᐧ

**Comment by algernon** [comment]
I'm fine with this plan.

**Comment by noseglasses** [comment]
Agreed. I will adapt this PR accordingly.

**Comment by noseglasses** [comment]
@algernon, what's the official deprecation period to be used in UPGRADING.md?

**Comment by algernon** [comment]
That's a good question... I've been pretty lax about adding those recently. A quick check suggests we used ~3 months as the default period.

**Comment by noseglasses** [comment]
> and the class currently called
Kaleidoscope inside the kaleidoscope namespace be renamed to "Runtime"
inside the kaleidoscope namespace, with back-compat left around for a bit.

Currently users call `Kaleidoscope.setup()` in their code. If we deprecate global instance `Kaleidoscope`, users would have to call `kaleidoscope::Runtime.setup()` instead.

We could leave the old deprecated `Kaleidoscope` symbol hang around forever to be on the safe side.

Is that what we want? 

I am just asking because then I would also update the stuff in `examples/...` to prevent us being flooded with deprecation messages on `make smoke-examples`.

**Comment by noseglasses** [comment]
Just force pushed a version that leaves the original `::Kaleidoscope` symbol untouched.

I opt for a transition to `kaleidoscope::Runtime` but can also live with a non-deprecated compatibility symbol `::Kaleidoscope` (2 bytes of RAM and some PROGMEM) instead.

**Comment by obra** [comment]
I think that leaving the toplevel Kaleidoscope symbol hanging around for
sketches to use makes sense, since it's "arduinoish"

Thanks for the updated PR
ᐧ

File: src/kaleidoscope/device/dygma/raise/SideFlash.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 #ifdef ARDUINO_SAMD_RAISE
 
-#include <Kaleidoscope.h>
+#include <Runtime.h>
```
Comment (obra) [inline src/kaleidoscope/device/dygma/raise/SideFlash.h]: Should all of these "#include <Runtime.h>" be "#include "kaleidoscope/Runtime.h" instead?

Comment (noseglasses) [reply src/kaleidoscope/device/dygma/raise/SideFlash.h]: Correct. The Raise was the only device that got not updated and still used the old service-include file.

Will fix that.

**Comment by obra** [comment]
(I thought I saw some other uses of "#include <Runtime.h>", too)
ᐧ

**Comment by noseglasses** [comment]
Alright, will check it thoroughly.

**Comment by noseglasses** [comment]
There were indeed several ocurrences of `#include <Runtime.h>` but all of them were in the Raise device directory subtree. I fixed them all.

I added another small change:
As we want `::Kaleidoscope` to be available to end users in their sketches, I moved its definition to `src/Kaleidoscope.h`. I did the same with the deprecated type definition `::Kaleidoscope_`. Both are now only available via header `src/Kaleidoscope.h` which we do not want to be included in other places than the sketch and 3rd party plugins.

On the other hand, I removed the type definition `::Runtime_` I had introduced in a previous version of this PR from `kaleidoscope/Runtime.h`. That's because we do not want anything that we use internally to be defined outside of namespace `kaleidoscope`.

What's your opinion?

**Comment by obra** [comment]
@noseglasses - that sounds reasonable :) I'm merging
